Textarea incorrectly reserves space for overlay scrollbar

lukewarlow opened the pull request:
#### cda20f66eecb111c71be0960b981e596473e889f
<pre>
Textarea incorrectly reserves space for overlay scrollbar
<a href="https://bugs.webkit.org/show_bug.cgi?id=275114">https://bugs.webkit.org/show_bug.cgi?id=275114</a>

Reviewed by Simon Fraser.

Check the ScrollbarTheme::usesOverlayScrollbar in the various ScrollbarTheme&apos;s scrollbarThickness method.
Update RenderTextControl to tell scrollbarThickness to ignore overlay scrollbar sizing.

This aligns WebKit with Firefox.

* LayoutTests/fast/scrolling/mac/scrollbars/overlay-scrollbar-textarea-cols-size-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/scrollbars/overlay-scrollbar-textarea-cols-size.html: Added.
* LayoutTests/fast/scrolling/mac/scrollbars/overlay-scrollbar-textarea-rows-size-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/scrollbars/overlay-scrollbar-textarea-rows-size.html: Added.
* Source/WebCore/platform/ScrollTypes.h:
* Source/WebCore/platform/ScrollbarTheme.h:
(WebCore::ScrollbarTheme::scrollbarThickness):
* Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.cpp:
(WebCore::ScrollbarThemeAdwaita::scrollbarThickness):
* Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.h:
* Source/WebCore/platform/gtk/ScrollbarThemeGtk.cpp:
(WebCore::ScrollbarThemeGtk::scrollbarThickness):
* Source/WebCore/platform/gtk/ScrollbarThemeGtk.h:
* Source/WebCore/platform/ios/ScrollbarThemeIOS.h:
* Source/WebCore/platform/ios/ScrollbarThemeIOS.mm:
(WebCore::ScrollbarThemeIOS::scrollbarThickness):
* Source/WebCore/platform/mac/ScrollbarThemeMac.h:
* Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
(WebCore::ScrollbarThemeMac::scrollbarThickness):
* Source/WebCore/platform/mock/ScrollbarThemeMock.cpp:
(WebCore::ScrollbarThemeMock::scrollbarThickness):
* Source/WebCore/platform/mock/ScrollbarThemeMock.h:
* Source/WebCore/platform/playstation/ScrollbarThemePlayStation.cpp:
(WebCore::ScrollbarThemePlayStation::scrollbarThickness):
* Source/WebCore/platform/playstation/ScrollbarThemePlayStation.h:
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderScrollbarTheme.h:
* Source/WebCore/rendering/RenderTextControl.cpp:
(WebCore::RenderTextControl::scrollbarThickness const):

Canonical link: <a href="https://commits.webkit.org/287837@main">https://commits.webkit.org/287837@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1893e453da77f5a0b319c3feed558fbf52513f73

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80706 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/228 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34642 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85233 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31690 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82817 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/245 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8025 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63037 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20827 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83775 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/91 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73463 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43339 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/255 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27623 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30147 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71589 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28148 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86665 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7933 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5608 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71331 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8110 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69299 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70572 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17625 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14589 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13535 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7895 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13416 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7734 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11253 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9540 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderTextControl.cpp:
```diff
@@ -87,7 +87,7 @@ void RenderTextControl::styleDidChange(StyleDifference diff, const RenderStyle*
 {
     // FIXME: We should get the size of the scrollbar from the RenderTheme instead.
-    return ScrollbarTheme::theme().scrollbarThickness(this->style().scrollbarWidth());
+    return ScrollbarTheme::theme().usesOverlayScrollbars() ? 0 : ScrollbarTheme::theme().scrollbarThickness(this->style().scrollbarWidth());
```
This is duplicating logic that `verticalScrollbarIntrusion()` already has. Can we share code? Or, maybe `ScrollbarTheme::theme().scrollbarThickness()` should take an argument to specify how the result should behave for overly scrollbars.

lukewarlow replied:
That should be fixed now. I pass an argument into scrollbarThickness()

smfr commented on LayoutTests/fast/scrolling/mac/scrollbars/overlay-scrollbar-textarea-cols-size.html:
```diff
@@ -0,0 +1,39 @@
+
+        async function doTest()
+        {
+            description('Test maybegin and cancelled');
```
Does it?

smfr commented on LayoutTests/fast/scrolling/mac/scrollbars/overlay-scrollbar-textarea-rows-size.html:
```diff
@@ -0,0 +1,39 @@
+
+        async function doTest()
+        {
+            description('Test maybegin and cancelled');
```
Ditto

smfr commented on LayoutTests/fast/scrolling/mac/scrollbars/overlay-scrollbar-textarea-rows-size.html:
```diff
@@ -0,0 +1,39 @@
+    <textarea cols="3" id="textareaA"></textarea>
+    <textarea cols="3" style="overflow: hidden" id="textareaB"></textarea>
+    <script src="../../../../resources/js-test-pre.js"></script>
+    <script src="../../../../resources/ui-helper.js"></script>
```
Can be removed.

smfr commented on LayoutTests/fast/scrolling/mac/scrollbars/overlay-scrollbar-textarea-rows-size.html:
```diff
@@ -0,0 +1,39 @@
+<!DOCTYPE html> <!-- webkit-test-runner [ MockScrollbarsEnabled=false AsyncOverflowScrollingEnabled=true ] -->
+<html>
+<head>
+    <textarea cols="3" id="textareaA"></textarea>
```
Why are your textareas in the `<head>` ?

smfr commented on LayoutTests/fast/scrolling/mac/scrollbars/overlay-scrollbar-textarea-cols-size.html:
```diff
@@ -0,0 +1,39 @@
+<!DOCTYPE html> <!-- webkit-test-runner [ MockScrollbarsEnabled=false AsyncOverflowScrollingEnabled=true ] -->
+<html>
+<head>
+    <textarea rows="3" id="textareaA" style="overflow-x: scroll"></textarea>
```
Why are your textareas in the `<head>`?

smfr commented on LayoutTests/fast/scrolling/mac/scrollbars/overlay-scrollbar-textarea-cols-size.html:
```diff
@@ -0,0 +1,39 @@
+    <textarea rows="3" id="textareaA" style="overflow-x: scroll"></textarea>
+    <textarea rows="3" style="overflow: hidden" id="textareaB"></textarea>
+    <script src="../../../../resources/js-test-pre.js"></script>
+    <script src="../../../../resources/ui-helper.js"></script>
```
Can be removed

smfr commented on Source/WebCore/platform/ScrollbarTheme.h:
```diff
@@ -40,6 +40,8 @@ class ScrollView;
 class GraphicsLayer;
 #endif
 
+enum OverlaySizeRelevancy : bool { Include, Ignore };
```
Can we use the existing `OverlayScrollbarSizeRelevancy` ?

lukewarlow committed:
Textarea incorrectly reserves space for overlay scrollbar
https://bugs.webkit.org/show_bug.cgi?id=275114

Reviewed by Simon Fraser.

Check the ScrollbarTheme::usesOverlayScrollbar in the various ScrollbarTheme's scrollbarThickness method.
Update RenderTextControl to tell scrollbarThickness to ignore overlay scrollbar sizing.

This aligns WebKit with Firefox.

* LayoutTests/fast/scrolling/mac/scrollbars/overlay-scrollbar-textarea-cols-size-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/scrollbars/overlay-scrollbar-textarea-cols-size.html: Added.
* LayoutTests/fast/scrolling/mac/scrollbars/overlay-scrollbar-textarea-rows-size-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/scrollbars/overlay-scrollbar-textarea-rows-size.html: Added.
* Source/WebCore/platform/ScrollTypes.h:
* Source/WebCore/platform/ScrollbarTheme.h:
(WebCore::ScrollbarTheme::scrollbarThickness):
* Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.cpp:
(WebCore::ScrollbarThemeAdwaita::scrollbarThickness):
* Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.h:
* Source/WebCore/platform/gtk/ScrollbarThemeGtk.cpp:
(WebCore::ScrollbarThemeGtk::scrollbarThickness):
* Source/WebCore/platform/gtk/ScrollbarThemeGtk.h:
* Source/WebCore/platform/ios/ScrollbarThemeIOS.h:
* Source/WebCore/platform/ios/ScrollbarThemeIOS.mm:
(WebCore::ScrollbarThemeIOS::scrollbarThickness):
* Source/WebCore/platform/mac/ScrollbarThemeMac.h:
* Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
(WebCore::ScrollbarThemeMac::scrollbarThickness):
* Source/WebCore/platform/mock/ScrollbarThemeMock.cpp:
(WebCore::ScrollbarThemeMock::scrollbarThickness):
* Source/WebCore/platform/mock/ScrollbarThemeMock.h:
* Source/WebCore/platform/playstation/ScrollbarThemePlayStation.cpp:
(WebCore::ScrollbarThemePlayStation::scrollbarThickness):
* Source/WebCore/platform/playstation/ScrollbarThemePlayStation.h:
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderScrollbarTheme.h:
* Source/WebCore/rendering/RenderTextControl.cpp:
(WebCore::RenderTextControl::scrollbarThickness const):

Canonical link: https://commits.webkit.org/287837@main
